router.go

```diff
@@ -36,19 +37,27 @@ type serviceProcedure struct {
 	procedure string
 }
 
+type serviceProcedureEncoding struct {
+	service   string
+	procedure string
+	encoding  transport.Encoding
+}
+
 // MapRouter is a Router that maintains a map of the registered
 // procedures.
 type MapRouter struct {
-	defaultService string
-	entries        map[serviceProcedure]transport.Procedure
+	defaultService            string
+	serviceProcedures         map[serviceProcedure]transport.Procedure
+	serviceProcedureEncodings map[serviceProcedureEncoding]transport.Procedure
```

Comment:
For the sake of consideration, we could put Register on the encodingProcedure implementation to handle the conflicting cases (where encoding is specified and a wildcard encoding is already in place, where an encoding is omitted where a specific encoding is already in place).

The downside is that it adds an interface indirection where it’s not necessary to facilitate polymorphism for the user. Ideally we can shape this code so removing support for `serviceProcedures` is strictly deleting some lines, not refactoring.